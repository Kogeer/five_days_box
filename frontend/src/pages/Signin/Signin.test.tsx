import {act, fireEvent, render, screen, waitFor} from "@testing-library/react";
import {MemoryRouter} from "react-router-dom";
import SignInPage from "./index";
import AuthStoreProvider from "../../components/Context/AuthContext";
import mock from "../../test/axios-mock";
import axiosInstance from "../../utils/axios";
import {setFieldValue} from "../../test/helper";

const mockHistoryPush = jest.fn();

jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useHistory: () => ({
        push: mockHistoryPush
    })
}));

const component = (
    <AuthStoreProvider>
        <MemoryRouter>
            <SignInPage/>
        </MemoryRouter>
    </AuthStoreProvider>
);

describe('Signin', () => {
    it('Render signin', async () => {
        const {asFragment} = render(component)

        expect(asFragment()).toMatchSnapshot();
    });

    it('Success signin', async () => {
        const spy = jest.spyOn(axiosInstance, 'post');
        mock.onGet('/auth/currentuser').reply(401, {});
        mock.onPost('/auth/signin').replyOnce(200, {data: {name: 'halo', id: 'mockid'}});

        const {getByPlaceholderText, getByRole} = render(component);

        const emailInput = getByPlaceholderText('E-mail');
        const passwordInput = getByPlaceholderText('Password');

        expect(emailInput).toBeInTheDocument();
        expect(passwordInput).toBeInTheDocument();

        await setFieldValue(emailInput, 'test@example.com')
        await setFieldValue(passwordInput, 'password');

        expect(emailInput).toHaveValue('test@example.com');
        expect(passwordInput).toHaveValue('password');
        const submit = getByRole('button', {name: 'Sign in'});

        expect(submit).toBeInTheDocument();

        await act(async () => {
            fireEvent.click(submit);
        });

        await waitFor(() => {
            expect(spy).toHaveBeenCalledWith('/auth/signin', {
                email: 'test@example.com',
                password: 'password'
            })
        });
        await waitFor(() => expect(mockHistoryPush).toHaveBeenCalledWith('/'));
    });
});
