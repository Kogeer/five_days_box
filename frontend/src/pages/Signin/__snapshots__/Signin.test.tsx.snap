// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Signin Render signin 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <h1
      class="text-center"
    >
      Sign in
    </h1>
    <div
      class="align-middle"
    >
      <form
        class=""
        novalidate=""
      >
        <div
          class="mb-3"
        >
          <input
            class="form-control"
            name="email"
            placeholder="E-mail"
          />
          <div
            class="invalid-feedback"
          >
            E-mail is required!
          </div>
        </div>
        <div
          class="mb-3"
        >
          <input
            class="form-control"
            name="password"
            placeholder="Password"
            type="password"
          />
          <div
            class="invalid-feedback"
          >
            Password is required and minimum 4 length!
          </div>
        </div>
        <div
          class="d-flex justify-content-between"
        >
          <button
            class="btn btn-primary"
            type="button"
          >
            Sign in
          </button>
          <a
            class="d-flex justify-content-center"
            href="/auth/signup"
          >
            <button
              class="btn btn-primary"
              type="button"
            >
              Sign up
            </button>
          </a>
        </div>
      </form>
    </div>
  </div>
</DocumentFragment>
`;
